lib/simulator_sse.h

```diff
@@ -0,0 +1,614 @@
+      rm = _mm_sub_ps(rm, _mm_mul_ps(i2, iu));
+      im = _mm_add_ps(im, _mm_mul_ps(i2, ru));
+      ru = _mm_set1_ps(matrix[30]);
+      iu = _mm_set1_ps(matrix[31]);
+      rm = _mm_add_ps(rm, _mm_mul_ps(r3, ru));
+      im = _mm_add_ps(im, _mm_mul_ps(r3, iu));
+      rm = _mm_sub_ps(rm, _mm_mul_ps(i3, iu));
+      im = _mm_add_ps(im, _mm_mul_ps(i3, ru));
+
+      switch (q0) {
+      case 0:
+        rm = _mm_shuffle_ps(rm, rm, 128);  // 10000000
+        im = _mm_shuffle_ps(im, im, 128);
+        rn = _mm_blend_ps(rn, rm, 10);  // 1010
+        in = _mm_blend_ps(in, im, 10);
+        break;
+      case 1:
+        rn = _mm_shuffle_ps(rn, rm, 68);  // 01000100
+        in = _mm_shuffle_ps(in, im, 68);
+        break;
+      default:
+        // Cannot reach here.
+        break;
+      }
+
+      p = si | sizej;
+      _mm_store_ps(rstate + p, rn);
+      _mm_store_ps(rstate + p + 4, in);
+    };
+
+    ParallelFor::Run(num_threads_, sizei / 8, f,
+                     sizej, mask0, mask1, q0, matrix, rstate);
+  }
+
+  // Applies a two-qubit gate for qubit0 = 0 and qubit1 = 1.
+  // Performs sparse matrix-vector multiplication.
+  // The inner loop (V_i = \sum_j M_ij V_j) is unrolled by hand.
+  void ApplyGate2LL(
+      unsigned q0, unsigned q1, const fp_type* matrix, State& state) const {
+    uint64_t sizei = uint64_t{1} << (num_qubits_ + 1);
+
+    fp_type* rstate = StateSpace::RawData(state);
+
+    auto f = [](unsigned n, unsigned m, uint64_t i, const fp_type* u,
+                fp_type* rstate) {
+      uint64_t p = 8 * i;
+
+      fp_type s0r = rstate[p + 0];
+      fp_type s1r = rstate[p + 1];
+      fp_type s2r = rstate[p + 2];
+      fp_type s3r = rstate[p + 3];
+      fp_type s0i = rstate[p + 4];
+      fp_type s1i = rstate[p + 5];
+      fp_type s2i = rstate[p + 6];
+      fp_type s3i = rstate[p + 7];
+
+      rstate[p + 0] = s0r * u[0] - s0i * u[1] + s1r * u[2] - s1i * u[3]
+          + s2r * u[4] - s2i * u[5] + s3r * u[6] - s3i * u[7];
+      rstate[p + 4] = s0r * u[1] + s0i * u[0] + s1r * u[3] + s1i * u[2]
+          + s2r * u[5] + s2i * u[4] + s3r * u[7] + s3i * u[6];
+      rstate[p + 1] = s0r * u[8] - s0i * u[9] + s1r * u[10] - s1i * u[11]
+          + s2r * u[12] - s2i * u[13] + s3r * u[14] - s3i * u[15];
+      rstate[p + 5] = s0r * u[9] + s0i * u[8] + s1r * u[11] + s1i * u[10]
+          + s2r * u[13] + s2i * u[12] + s3r * u[15] + s3i * u[14];
+      rstate[p + 2] = s0r * u[16] - s0i * u[17] + s1r * u[18] - s1i * u[19]
+          + s2r * u[20] - s2i * u[21] + s3r * u[22] - s3i * u[23];
+      rstate[p + 6] = s0r * u[17] + s0i * u[16] + s1r * u[19] + s1i * u[18]
+          + s2r * u[21] + s2i * u[20] + s3r * u[23] + s3i * u[22];
+      rstate[p + 3] = s0r * u[24] - s0i * u[25] + s1r * u[26] - s1i * u[27]
+          + s2r * u[28] - s2i * u[29] + s3r * u[30] - s3i * u[31];
+      rstate[p + 7] = s0r * u[25] + s0i * u[24] + s1r * u[27] + s1i * u[26]
+          + s2r * u[29] + s2i * u[28] + s3r * u[31] + s3i * u[30];
```

Comment:
Why no intrinsics here ?

Reply:
Thank you for your review. This code is for qubit0=0 and qubit1=1. The memory layout is such that it's not vectorizable in this case.